src/lib/y2storage/storage_class_wrapper.rb

```diff
@@ -0,0 +1,258 @@
+      #     end
+      #   end
+      #
+      #   device = Y2Storage::BlkDevice.new(storage_device)
+      #
+      #   # This is simply forwarded
+      #   device.name
+      #
+      #   # Returns a Y2Storage::DiskSize object instead of the Fixnum (bytes)
+      #   # returned by Storage::BlkDevice#size
+      #   device.size
+      #
+      #   # Transforms the argument to Fixnum (bytes) before passing it to
+      #   # Storage::BlkDevice#size=
+      #   device.size = Y2Storage::DiskSize.GiB(20)
+      #
+      #   # Returns an array of strings instead of a Storage::VectorString
+      #   device.udev_paths
+      #
+      #   # If there is no filesystem (device is not formatted), it returns
+      #   # nil instead of raising an exception. If there is one, its wrapped
+      #   # into a Y2Storage::Filesystems::BlkFilesystem object before being
+      #   # returned.
+      #   device.blk_filesystem
+      #
+      #   fs_type = Y2Storage::Filesystems::Type.find(:btrfs)
+      #   # Transforms fs_type into something understandable by
+      #   # Storage::BlkDevice#create_blk_device when passing it.
+      #   # If the device was already formatted, it raises an exception
+      #   # (is not masked like in the example above).
+      #   device.create_blk_filesystem(fs_type)
+      #
+      def storage_forward(method, as: nil, raise_errors: [])
+        modifiers = { as: as, raise_errors: raise_errors }
+        define_method(method) do |*args|
+          StorageClassWrapper.forward(to_storage_value, method, modifiers, *args)
+        end
+      end
+
+      # Equivalent to #storage_forward used to define class methods instead of
+      # instance ones.
+      #
+      # @see #storage_forward
+      def storage_class_forward(method, as: nil, raise_errors: [])
+        modifiers = { as: as, raise_errors: raise_errors }
+        define_singleton_method(method) do |*args|
+          StorageClassWrapper.forward(storage_class, method, modifiers, *args)
+        end
+      end
+
+      def storage_class
+        @storage_class
+      end
+
+      def storage_class_name
+        @storage_class_name ||= storage_class.name.split("::").last
+      end
+
+      def storage_class_underscored_name
+        @storage_class_underscored_name ||= StorageClassWrapper.underscore(storage_class_name)
+      end
+
+      # Alternative constructor used internally by this module in order to get
+      # fully downcasted objects.
+      #
+      # @param storage object to be wrapped
+      def downcasted_new(object)
+        @downcast_class_names.each do |class_name|
+          underscored = StorageClassWrapper.underscore(class_name.split("::").last)
+          check_method = :"#{underscored}?"
+          cast_method = :"to_#{underscored}"
+          next unless Storage.send(check_method, object)
+
+          klass = StorageClassWrapper.class_for(class_name)
+          return klass.downcasted_new(Storage.send(cast_method, object))
```

Comment:
bumm, class_for can return nil. In general I suggest to raise error if class_for does not find class.